src/Nethermind/Nethermind.Consensus/Processing/RecoverSignature.cs

```diff
@@ -45,21 +45,55 @@ public void RecoverData(Block block)
                 // so we assume the rest of txs in the block are already recovered
                 return;
 
-            var releaseSpec = _specProvider.GetSpec(block.Header);
-
             Parallel.ForEach(
-                block.Transactions.Where(tx => tx.IsSigned && tx.SenderAddress is null),
+                block.Transactions.Where(tx => !tx.IsHashCalculated),
                 blockTransaction =>
                 {
-                    _txPool.TryGetPendingTransaction(blockTransaction.Hash, out Transaction? transaction);
+                    blockTransaction.CalculateHashInternal();
+                });
 
-                    Address sender = transaction?.SenderAddress;
-                    Address blockTransactionAddress = blockTransaction.SenderAddress;
+            var releaseSpec = _specProvider.GetSpec(block.Header);
 
-                    blockTransaction.SenderAddress =
-                        sender ?? _ecdsa.RecoverAddress(blockTransaction, !releaseSpec.ValidateChainId);
-                    if (_logger.IsTrace) _logger.Trace($"Recovered {blockTransaction.SenderAddress} sender for {blockTransaction.Hash} (tx pool cached value: {sender}, block transaction address: {blockTransactionAddress})");
-                });
+            int recoverFromEcdsa = 0;
+            // Don't access txPool in Parallel loop as increases contention
+            foreach (Transaction blockTransaction in block.Transactions.Where(tx => tx.IsSigned && tx.SenderAddress is null))
+            {
+                _txPool.TryGetPendingTransaction(blockTransaction.Hash, out Transaction? transaction);
+
+                Address sender = transaction?.SenderAddress;
+                if (sender != null)
+                {
+                    blockTransaction.SenderAddress = sender;
+
+                    if (_logger.IsTrace) _logger.Trace($"Recovered {blockTransaction.SenderAddress} sender for {blockTransaction.Hash} (tx pool cached value: {sender})");
+                }
+                else
+                {
+                    recoverFromEcdsa++;
+                }
+            }
+
+            if (recoverFromEcdsa >= 4)
```

Comment:
why this threshold? ECDSA recovery should be slow enough that it is always good to parallelize?

Reply:
1 always runs inline; should should be greater than that; then is overhead of additional allocations from `Parallel` and it's set up; so chose 4 which is really more than 3 to do in parallel.

From tracing the majority take the parallel path but some do take the serial path so is valid